browser/src/ConstraintTable/ConstraintTable.spec.tsx

```diff
@@ -0,0 +1,109 @@
+import React from 'react'
+import { Factory } from 'fishery' // eslint-disable-line import/no-extraneous-dependencies
+
+import renderer from 'react-test-renderer'
+import { createRenderer as createShallowRenderer } from 'react-test-renderer/shallow'
+import { forAllDatasets, forAllDatasetsExcept } from '../../../tests/__helpers__/datasets'
+
+import ConstraintTable, { Gene, Transcript } from './ConstraintTable'
+
+const geneFactory = Factory.define<Gene>(() => ({
+  chrom: '13',
+  transcripts: [],
+}))
+
+const transcriptFactory = Factory.define<Transcript>(() => ({
+  transcript_id: 'dummy_transcript',
+  transcript_version: '12.34.5',
+  chrom: '13',
+}))
+
+forAllDatasets('ConstraintTable with "%s" dataset selected', (datasetId) => {
+  describe('with a minimal gene', () => {
+    test('has no unexpected changes', () => {
+      const tree = renderer.create(
+        <ConstraintTable datasetId={datasetId} geneOrTranscript={geneFactory.build()} />
+      )
+      expect(tree).toMatchSnapshot()
+    })
+  })
+
+  describe('with a minimal transcript', () => {
+    test('has no unexpected changes', () => {
+      const tree = renderer.create(
+        <ConstraintTable datasetId={datasetId} geneOrTranscript={transcriptFactory.build()} />
+      )
+      expect(tree).toMatchSnapshot()
+    })
+  })
+
+  describe('with a mitochondrial gene', () => {
+    test('has no unexpected changes', () => {
+      const tree = renderer.create(
+        <ConstraintTable
+          datasetId={datasetId}
+          geneOrTranscript={geneFactory.build({ chrom: 'M' })}
+        />
+      )
+      expect(tree).toMatchSnapshot()
+    })
+  })
+
+  describe('with a mitochondrial transcript', () => {
+    test('has no unexpected changes', () => {
+      const tree = renderer.create(
+        <ConstraintTable
+          datasetId={datasetId}
+          geneOrTranscript={transcriptFactory.build({ chrom: 'M' })}
+        />
+      )
+      expect(tree).toMatchSnapshot()
+    })
+  })
+})
+
+test('ConstraintTable with exac dataset and gene with available constraints has no unexpected changes', () => {
+  const tree = createShallowRenderer().render(
```

Comment:
More just curiosity - what's the reasoning behind when to use `createShallowRenderer()`? It looks like the exac table has less other components in it, and the use case seems to be for testing of components in isolation, regardless of how many other components they are built up of? So I would've expected `createShallowRenderer()` to be used either everywhere, or more heavily for datasets that aren't `exac`. I may just ask in zoom, but am curious.

Reply:
This might just be a mistake on my part, I'm looking into that.

Reply:
@rileyhgrant `createShallowRenderer` is in fact not appropriate here, and in moving away from that, I saw another chunk of refactoring that I want to include in this PR, so watch this space for a re-review request.

The basic difference between shallow and regular rendering is that regular rendering recursively renders all children of the rendered node, until it doesn't have any React components in it, just HTML elements, text, other raw elements of a Web page. By contrast, shallow rendering doesn't render the children: instead of rendering a child, it includes a kind of dummy node indicating what child was called and with what params.

One common use case for shallow rendering would be something like this: say you have a `List` widget that contains some number of children, each of which is a heavyweight React component in its own right, with maybe a bunch of network calls that have to be stubbed, test data that has to be set just right, and other complications to the test setup. Suppose you're writing a test to verify that, if you render a `List` with the right arguments, you get a certain sequence of child widgets--the content itself of the widgets is of no interest to you in this one particular test, you just want to know that each is the right kind called with the correct set of arguments. Here shallow rendering lets you test what you're concerned about without having to do a lot of extra setup just to ensure that the parts you don't really care about don't fail the test.